File: Data Structures & Algorithms/unique-paths-ii/submission-1.java
class Solution {
    int[][] cache;

    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        cache = new int[obstacleGrid.length][obstacleGrid[0].length];
        return dp(obstacleGrid, 0, 0);
    }

    private int dp(int[][] grid, int r, int c) {
        if (r == grid.length || c == grid[0].length || grid[r][c] == 1) {
            return 0;
        }

        if (cache[r][c] > 0) {
            return cache[r][c];
        }

        if (r == grid.length - 1 && c == grid[0].length - 1) {
            return 1;
        }

        cache[r][c] = dp(grid, r + 1, c) + dp(grid, r, c + 1);

        return cache[r][c];
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.kt
/*
// Definition for a Node.
class Node(var `val`: Int) {
    var next: Node? = null
    var random: Node? = null
}
*/

class Solution {
    fun copyRandomList(head: Node?): Node? {
        val nodeToCopy = mutableMapOf<Node, Node>()
        return recursiveCopy(head, nodeToCopy)
    }

    private fun recursiveCopy(node: Node?, cache: MutableMap<Node, Node>): Node? {
        if (node == null) {
            return null
        }

        if (node in cache) {
            return cache[node]
        }

        val copy = Node(node.`val`)
        cache[node] = copy
        copy.next = recursiveCopy(node.next, cache)
        copy.random = recursiveCopy(node.random, cache)

        return cache[node]
    }
}
